prow/crier/reporters/github/reporter.go

```diff
@@ -35,6 +39,21 @@ type Client struct {
 	gc          report.GitHubClient
 	config      config.Getter
 	reportAgent v1.ProwJobAgent
+	prLocks     *shardedLock
+}
+
+type shardedLock struct {
+	mapLock *sync.Mutex
+	locks   map[string]*sync.Mutex
+}
+
+func (s *shardedLock) getLock(key string) *sync.Mutex {
```

Comment:
I should probably mention that this is a memory leak, as we allocate a lock for each unique PR for which we observe a job and can never clean it up again. I would think though that this is not going to allocate so much memory that this will be a problem. WDYT?

Reply:
Under the shower I realized that we can actually free them, by just acquiring the map lock, then acquiring the locks themselves, then deleting them. This essentially prevents all presubmit reporting from happening so shouldn't be done too often, but no reason to not do that.

I'll fix up the PR to do that once I find some time.

Reply:
SGTM!

Reply:
Added a cleanup for the locks